llm-models/embedding/automl/automl_embeddings_example.py

```diff
@@ -0,0 +1,147 @@
+
+# COMMAND ----------
+
+import mlflow
+from langchain.embeddings.base import Embeddings
+from typing import Any, Dict, List, Optional
+
+# Creating a version of Langchain's Embeddings that can support a mlflow experiment's run_id
+class SentenceTransformerFromMlflow(Embeddings):
+    def __init__(self, run_id: str):
+        # Gets the latest model trained by AutoML
+        latest_model_path = mlflow.MlflowClient().list_artifacts(run_id)[-1].path
+        model_uri = f"runs:/{run_id}/{latest_model_path}"
+        self.model = mlflow.sentence_transformers.load_model(model_uri)
+
+    def embed_documents(self, texts: List[str]) -> List[List[float]]:
+        # Prefix each text with "passage: " if they aren't already prefixed
+        texts = [text if text.startswith("passage:") else f"passage: {text}" for text in texts]
+        embeddings = self.model.encode(texts)
+        return embeddings.tolist()
+
+    def embed_query(self, text: str) -> List[float]:
+        # Prefix query with "query: " if it isn't already prefixed
+        text = text if text.startswith("query:") else f"query: {text}"
+        embedding = self.model.encode(text)
+        return embedding.tolist()
+
+# COMMAND ----------
+
+# MAGIC %md ### Querying the FAISS Vector Store
+# MAGIC
+# MAGIC This is a simple example of creating a FAISS vector store using an existing Spark DataFrame using the [PySparkDataFrameLoader](https://python.langchain.com/docs/integrations/document_loaders/pyspark_dataframe)
+
+# COMMAND ----------
+
+from langchain.vectorstores import FAISS
+from langchain.document_loaders import PySparkDataFrameLoader
+
+fine_tuned_embeddings_model = SentenceTransformerFromMlflow(run_id = run_id)
+spark_df = spark.table(delta_table_name)
+loader = PySparkDataFrameLoader(spark, spark_df, page_content_column=text_column_name)
+documents = loader.load()
+
+faiss_vector_store = FAISS.from_documents(documents, embedding=fine_tuned_embeddings_model)
+
+# COMMAND ----------
+
+faiss_vector_store.similarity_search("what is the delta engine")[0]
+
+# COMMAND ----------
+
+# MAGIC %md ### Using an Existing Model for Comparison
+# MAGIC
+# MAGIC Here, we will use the model `intfloat/e5-base-v2` ([source](https://huggingface.co/intfloat/e5-base-v2)) from the [MTEB](https://huggingface.co/spaces/mteb/leaderboard) benchmark to compare with our fine-tuned model.
+
+# COMMAND ----------
+
+from langchain.embeddings import HuggingFaceEmbeddings
+
+sota_embeddings_model = HuggingFaceEmbeddings(model_name="intfloat/e5-base-v2")
+sota_faiss_vector_store = FAISS.from_documents(documents, embedding=sota_embeddings_model)
+
+# COMMAND ----------
+
+sota_faiss_vector_store.similarity_search("query: what is the delta engine")[0]
+
+# COMMAND ----------
+
+# MAGIC %md ### (Optional) Creating a Doc Q&A Bot using OpenAI
```

Comment:
There isn't really a simple way for a customer to be able to run a Q&A bot unless they have OpenAI, so I made this cell optional for now, but any other suggestions are welcome!

Reply:
Explicitly marking it as optional sounds good to me 👍